File: AgoraChat/Code/Controllers/Threads/AgoraChatThreadListViewController.m
//
//  AgoraChatThreadListViewController.m
//  AgoraChat
//
//  Created by 朱继超 on 2022/4/5.
//  Copyright © 2022 easemob. All rights reserved.
//

#import "AgoraChatThreadListViewController.h"
#import "AgoraChatThreadListNavgation.h"
#import "AgoraChatUserDataModel.h"
#import "UserInfoStore.h"
#import "AgoraChatThreadViewController.h"
@interface AgoraChatThreadListViewController ()<EaseThreadListProtocol>

@property (nonatomic) AgoraChatThreadListNavgation *navBar;

@property (nonatomic) EaseThreadListViewController *chatController;


@end

@implementation AgoraChatThreadListViewController

- (instancetype)initWithGroup:(AgoraChatGroup *)group chatViewModel:(EaseChatViewModel *)viewModel {
    if ([super init]) {
        self.chatController = [[EaseThreadListViewController alloc]initWithGroup:group chatViewModel:viewModel];
        self.chatController.delegate = self;
    }
    return self;
}

- (void)threadListCount:(int)count {
    self.navBar.title = [NSString stringWithFormat:@"All threads (%d)",count];
}

- (void)agoraChatThreadList:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    EaseThreadConversation *conv = self.chatController.dataArray[indexPath.row];
    [AgoraChatClient.sharedClient.threadManager joinChatThread:conv.threadInfo.threadId completion:^(AgoraChatThread *thread, AgoraChatError *aError) {
        if (!aError || aError.code == 40004) {
            AgoraChatThreadViewController *VC = [[AgoraChatThreadViewController alloc]initThreadChatViewControllerWithCoversationid:conv.threadInfo.threadId conversationType:AgoraChatConversationTypeGroupChat chatViewModel:self.chatController.viewModel parentMessageId:@"" model:nil];
            VC.navTitle = thread ? thread.threadName:conv.threadInfo.threadName;
            VC.detail = self.chatController.group.groupName;
            [self.navigationController pushViewController:VC animated:YES];
        }
    }];

}
- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.navigationController.navigationBarHidden = YES;
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    self.navigationController.navigationBarHidden = NO;
}


- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self _setupChatSubviews];
}

- (AgoraChatThreadListNavgation *)navBar {
    if (!_navBar) {
        _navBar = [[AgoraChatThreadListNavgation alloc]initWithFrame:CGRectMake(0, 0, EMScreenWidth, EMNavgationHeight)];
        _navBar.backgroundColor = [UIColor colorWithHexString:@"#FFFFFF"];
        __weak typeof(self) weakSelf = self;
        [_navBar setBackBlock:^{
            [weakSelf.navigationController popViewControllerAnimated:YES];
        }];
    }
    return _navBar;
}

- (void)_setupChatSubviews
{
    [self addChildViewController:self.chatController];
    [self.view addSubview:self.navBar];
    self.navBar.title = @"All Threads";
    self.navBar.detail = [NSString stringWithFormat:@"# %@",self.chatController.group.groupName];
    [self.view addSubview:self.chatController.view];
    [self.chatController.view mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view).insets(UIEdgeInsetsMake(EMNavgationHeight, 0, 0, 0));
    }];
    self.view.backgroundColor = [UIColor colorWithRed:242/255.0 green:242/255.0 blue:242/255.0 alpha:1.0];
}

//userProfile
- (id<EaseUserProfile>)userProfile:(NSString *)huanxinID
{
    AgoraChatUserDataModel *model = nil;
    if ([huanxinID isEqualToString:@""] || huanxinID == nil) {
        return model;
    }
    AgoraChatUserInfo* userInfo = [[UserInfoStore sharedInstance] getUserInfoById:huanxinID];
    if(userInfo) {
        model = [[AgoraChatUserDataModel alloc]initWithUserInfo:userInfo];
    }else{
        [[UserInfoStore sharedInstance] fetchUserInfosFromServer:@[huanxinID]];
    }
    return model;
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: AgoraChat-Swift/AgoraChat-Swift/code/Controllers/Threads/AgoraChatThreadListViewController.swift
//
//  AgoraChatThreadListViewController.swift
//  AgoraChat-Swift
//
//  Created by 冯钊 on 2022/10/25.
//

import UIKit
import SnapKit

class AgoraChatThreadListViewController: UIViewController {

    private let navBar = AgoraChatThreadListNavgation()
    private let chatController: EaseThreadListViewController
    
    init(group: AgoraChatGroup, viewModel: EaseChatViewModel) {
        self.chatController = EaseThreadListViewController(group: group, chatViewModel: viewModel)
        super.init(nibName: nil, bundle: nil)
        self.chatController.delegate = self
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = Color_F2F2F2
        
        self.navBar.backgroundColor = .white
        self.navBar.backHandle = { [unowned self] in
            self.navigationController?.popViewController(animated: true)
        }
        self.navBar.title = "All Threads".localized
        self.view.addSubview(self.navBar)
        
        self.addChild(self.chatController)
        self.view.addSubview(self.chatController.view)
        
        self.chatController.view.snp.makeConstraints { make in
            make.left.right.bottom.equalTo(self.view)
            make.top.equalTo(self.view.safeAreaLayoutGuide).offset(44)
        }
        self.navBar.snp.makeConstraints { make in
            make.left.right.top.equalTo(self.view)
            make.bottom.equalTo(self.chatController.view.snp.top)
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: true)
    }
}

extension AgoraChatThreadListViewController: EaseThreadListProtocol {
    func agoraChatThreadList(_ tableView: UITableView!, didSelectRowAt indexPath: IndexPath!) {
        guard let conversation = self.chatController.dataArray[indexPath.row] as? EaseThreadConversation else {
            return
        }
        let message = AgoraChatClient.shared().chatManager?.getMessageWithMessageId(conversation.threadInfo.messageId)
        let model = EaseMessageModel()
        var haveMessage = false
        if let message = message, message.messageId.count > 0 {
            haveMessage = true
            model.message = message
            model.direction = message.direction
            model.type = AgoraChatMessageType(rawValue: message.body.type.rawValue)!
            model.isHeader = true
        }
        
        AgoraChatClient.shared().threadManager?.joinChatThread(conversation.threadInfo.threadId, completion: { thread, error in
            if error == nil || error?.code == .userAlreadyExist {
                if let thread = thread {
                    model.thread = thread
                }
                let vc = AgoraChatThreadViewController(conversationId: conversation.threadInfo.threadId, type: .groupChat, viewModel: self.chatController.viewModel, parentMessageId: message?.messageId ?? "", model: haveMessage ? model : nil)
                vc.navTitle = thread?.threadName ?? conversation.threadInfo.threadName
                vc.detail = self.chatController.group.groupName
                self.navigationController?.pushViewController(vc, animated: true)
            }
        })
    }
}
